module/VuFind/src/VuFind/RecordDriver/MarcReader.php

```diff
@@ -0,0 +1,495 @@
+<?php
+/**
+ * MARC record reader class.
+ *
+ * PHP version 7
+ *
+ * Copyright (C) The National Library of Finland 2020.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  RecordDrivers
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:record_drivers Wiki
+ */
+namespace VuFind\RecordDriver;
+
+/**
+ * MARC record reader class.
+ *
+ * @category VuFind
+ * @package  RecordDrivers
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:record_drivers Wiki
+ */
+class MarcReader
```

Comment:
There's a lot going on in this one class -- it reads and writes two different formats, plus it allows data extraction. I think it might be beneficial to consider abstracting the format-specific logic to separate classes so that it's easier to add additional encodings (such as MARC-in-JSON) as needed and so that it's easier to test discrete units of functionality. I have a few broad ideas of how we might be able to do that, but I haven't thought about it in great depth yet since I realize there may be performance trade-offs to consider. Let me know if you think this is worth discussing in greater depth!

Reply:
I'll move out the format specific stuff. Can still be refined as needed.